java/jni/YGJNI.cpp

```diff
@@ -24,54 +24,74 @@ static void YGTransferLayoutDirection(YGNodeRef node, alias_ref<jobject> javaNod
-    obj->setFieldValue(topField, YGNodeLayoutGetTop(root));
-
-    obj->setFieldValue(marginLeftField, YGNodeLayoutGetMargin(root, YGEdgeLeft));
-    obj->setFieldValue(marginTopField, YGNodeLayoutGetMargin(root, YGEdgeTop));
-    obj->setFieldValue(marginRightField, YGNodeLayoutGetMargin(root, YGEdgeRight));
-    obj->setFieldValue(marginBottomField, YGNodeLayoutGetMargin(root, YGEdgeBottom));
-
-    obj->setFieldValue(paddingLeftField, YGNodeLayoutGetPadding(root, YGEdgeLeft));
-    obj->setFieldValue(paddingTopField, YGNodeLayoutGetPadding(root, YGEdgeTop));
-    obj->setFieldValue(paddingRightField, YGNodeLayoutGetPadding(root, YGEdgeRight));
-    obj->setFieldValue(paddingBottomField, YGNodeLayoutGetPadding(root, YGEdgeBottom));
-
-    obj->setFieldValue(borderLeftField, YGNodeLayoutGetBorder(root, YGEdgeLeft));
-    obj->setFieldValue(borderTopField, YGNodeLayoutGetBorder(root, YGEdgeTop));
-    obj->setFieldValue(borderRightField, YGNodeLayoutGetBorder(root, YGEdgeRight));
-    obj->setFieldValue(borderBottomField, YGNodeLayoutGetBorder(root, YGEdgeBottom));
-
-    YGTransferLayoutDirection(root, obj);
-
-    for (uint32_t i = 0; i < YGNodeGetChildCount(root); i++) {
-      YGTransferLayoutOutputsRecursive(YGNodeGetChild(root, i));
+  if(YGNodeGetHasNewLayout(root)){
+    if (auto obj = YGNodeJobject(root)->lockLocal()) {
+      static auto widthField = obj->getClass()->getField<jfloat>("mWidth");
+      static auto heightField = obj->getClass()->getField<jfloat>("mHeight");
+      static auto leftField = obj->getClass()->getField<jfloat>("mLeft");
+      static auto topField = obj->getClass()->getField<jfloat>("mTop");
+
+      static auto marginLeftField = obj->getClass()->getField<jfloat>("mMarginLeft");
+      static auto marginTopField = obj->getClass()->getField<jfloat>("mMarginTop");
+      static auto marginRightField = obj->getClass()->getField<jfloat>("mMarginRight");
+      static auto marginBottomField = obj->getClass()->getField<jfloat>("mMarginBottom");
+
+      static auto paddingLeftField = obj->getClass()->getField<jfloat>("mPaddingLeft");
+      static auto paddingTopField = obj->getClass()->getField<jfloat>("mPaddingTop");
+      static auto paddingRightField = obj->getClass()->getField<jfloat>("mPaddingRight");
+      static auto paddingBottomField = obj->getClass()->getField<jfloat>("mPaddingBottom");
+
+      static auto borderLeftField = obj->getClass()->getField<jfloat>("mBorderLeft");
+      static auto borderTopField = obj->getClass()->getField<jfloat>("mBorderTop");
+      static auto borderRightField = obj->getClass()->getField<jfloat>("mBorderRight");
+      static auto borderBottomField = obj->getClass()->getField<jfloat>("mBorderBottom");
+
+      static auto hasNewLayoutField = obj->getClass()->getField<jboolean>("mHasNewLayout");
+      static auto edgeSetFlagField = obj->getClass()->getField<jint>("mEdgeSetFlag");
+
+      const int MARGIN = 1;
```

Comment:
comment that these need to be kept in sync with YogaNode.java (or pull them off that class)